src/sys/socket/addr.rs

```diff
@@ -792,58 +800,86 @@ impl SockAddr {
+    /// allocated and valid.  It must be at least as large as all the useful parts
+    /// of the structure.  Note that in the case of a `sockaddr_un`, `len` need not
+    /// include the terminating null.
     ///
     /// # Safety
-    ///
-    /// unsafe because it takes a raw pointer as argument.  The caller must
-    /// ensure that the pointer is valid.
-    #[cfg(not(target_os = "fuchsia"))]
-    pub unsafe fn from_raw_sockaddr(addr: *const libc::sockaddr) -> Option<SockAddr> {
-        if addr.is_null() {
-            None
-        } else {
-            match AddressFamily::from_i32(i32::from((*addr).sa_family)) {
-                Some(AddressFamily::Unix) => None,
-                Some(AddressFamily::Inet) => Some(SockAddr::Inet(
-                    InetAddr::V4(*(addr as *const libc::sockaddr_in)))),
-                Some(AddressFamily::Inet6) => Some(SockAddr::Inet(
-                    InetAddr::V6(*(addr as *const libc::sockaddr_in6)))),
-                #[cfg(any(target_os = "android", target_os = "linux"))]
-                Some(AddressFamily::Netlink) => Some(SockAddr::Netlink(
-                    NetlinkAddr(*(addr as *const libc::sockaddr_nl)))),
-                #[cfg(any(target_os = "ios", target_os = "macos"))]
-                Some(AddressFamily::System) => Some(SockAddr::SysControl(
-                    SysControlAddr(*(addr as *const libc::sockaddr_ctl)))),
-                #[cfg(any(target_os = "android", target_os = "linux"))]
-                Some(AddressFamily::Packet) => Some(SockAddr::Link(
-                    LinkAddr(*(addr as *const libc::sockaddr_ll)))),
-                #[cfg(any(target_os = "dragonfly",
-                          target_os = "freebsd",
-                          target_os = "ios",
-                          target_os = "macos",
-                          target_os = "netbsd",
-                          target_os = "illumos",
-                          target_os = "openbsd"))]
-                Some(AddressFamily::Link) => {
-                    let ether_addr = LinkAddr(*(addr as *const libc::sockaddr_dl));
-                    if ether_addr.is_empty() {
-                        None
-                    } else {
-                        Some(SockAddr::Link(ether_addr))
-                    }
-                },
-                #[cfg(any(target_os = "android", target_os = "linux"))]
-                Some(AddressFamily::Vsock) => Some(SockAddr::Vsock(
-                    VsockAddr(*(addr as *const libc::sockaddr_vm)))),
-                // Other address families are currently not supported and simply yield a None
-                // entry instead of a proper conversion to a `SockAddr`.
-                Some(_) | None => None,
+    /// `addr` must be a valid, non-null pointer, and `len` should describe the
+    /// number of bytes within `*addr` that are initialized and represent data.
+    pub unsafe fn from_raw_sockaddr(addr: *const libc::sockaddr, len: usize) -> Result<SockAddr> {
```

Comment:
Why add the `len` argument?  `struct sockaddr` isn't very big, unlike `sockaddr_storage`.  Are there ever any cases when the caller actually gets a partial `sockaddr`?  The case in `from_libc_ifaddrs` isn't a good example, because it determines the `len` strictly based on the contents of the sockaddr.

Reply:
`*const sockaddr` isn't actually used as "I am passing a direct reference to a `struct sockaddr`", it's used (paired with a len) as a general type to say "there is a sockaddr stored here of length `len`". In the bind(2) linux man page:

> The actual structure passed for the addr argument will depend on the address family.  The sockaddr structure is defined as something like:
> ```
> struct sockaddr {
>     sa_family_t sa_family;
>     char        sa_data[14];
> }
> ```
>
> The only purpose of this structure is to cast the structure pointer passed in addr in order to avoid compiler warnings.  See EXAMPLES below.

For example, bind() accepts a `const struct sockaddr *addr, socklen_t addrlen`, but that doesn't mean that the biggest address you can pass to it is 14 bytes long - you can definitely do `struct sockaddr_un u = ...; bind(fd, &u, sizeof(u))`. We do the same thing here with accepting a `*const sockaddr, usize` pair, just with turning it into a Rust type instead of naming the socket.